src/python/WMCore/ReqMgr/CherryPyThreads/StepChainParentageFixTask.py

```diff
@@ -51,52 +47,57 @@ def fixStepChainParentage(self, config):
         Fix the StepChain parentage and update the ParentageResolved flag to True
         """
         self.logger.info("Running fixStepChainParentage thread for statuses: %s", self.statusToCheck)
-        childDatasets = set()
-        requests = set()
-        requestsByChildDataset = {}
+        fixedCount = 0
         for status in self.statusToCheck:
-            reqByChildDS = getChildDatasetsForStepChainMissingParent(self.reqmgrDB, status)
-            self.logger.info("Retrieved %d datasets to fix parentage, in status: %s",
-                             len(reqByChildDS), status)
-            childDatasets = childDatasets.union(set(reqByChildDS.keys()))
-            # We need to just get one of the StepChain workflow if multiple workflow contains the same datasets. (i.e. ACDC)
-            requestsByChildDataset.update(reqByChildDS)
-
-            for wfs in viewvalues(reqByChildDS):
-                requests = requests.union(wfs)
-
-        failedRequests = set()
-        totalChildDS = len(childDatasets)
-        fixCount = 0
-        for childDS in childDatasets:
-            self.logger.info("Resolving parentage for dataset: %s", childDS)
-            start = time.time()
-            try:
-                failedBlocks = self.dbsSvc.fixMissingParentageDatasets(childDS, insertFlag=True)
-            except Exception as exc:
-                self.logger.exception("Failed to resolve parentage data for dataset: %s. Error: %s",
-                                      childDS, str(exc))
-                failedRequests = failedRequests.union(requestsByChildDataset[childDS])
-            else:
-                if failedBlocks:
-                    self.logger.warning("These blocks failed to be resolved and will be retried later: %s",
-                                        failedBlocks)
-                    failedRequests = failedRequests.union(requestsByChildDataset[childDS])
+            listWflowChildren = getChildDatasetsForStepChainMissingParent(self.reqmgrDB, status)
+            msg = f"Retrieved {len(listWflowChildren)} workflows to fix parentage in status {status}"
+            self.logger.info(msg)
+
+            # now resolve every dataset parentage for each workflow
+            while listWflowChildren:
+                wflowData = listWflowChildren.pop()
+                msg = f'Resolving parentage for workflow {wflowData["workflowName"]} '
+                msg += f'with a total of {wflowData["childrenDsets"]} datasets. '
+                msg += f'A total of {len(listWflowChildren)} workflows still to be resolved.'
+                self.logger.info(msg)
+                # measure time taken to fix a workflow parentage
+                start = time.time()
+                parentageResolved = True
+                for childDS in wflowData["childrenDsets"]:
```

Comment:
again there is no check for `wflowData` dict here, I rather prefer to change this line to
```
for childDS in wflowData.get('childredDsets', []):
```

Reply:
This is already accounted and there is no need for double default. Please see `getChildDatasetsForStepChainMissingParent` where it is already set to a default `[]`, in case no datasets exist.